tools/build.py

```diff
@@ -0,0 +1,681 @@
+
+
+def clean_targets(cmake_path, build_dir: Path, configs: Set[str]):
+    for config in configs:
+        log.info("Cleaning targets for %s configuration", config)
+        build_dir2 = _get_build_config_dir(build_dir, config)
+        cmd_args = [cmake_path, "--build", build_dir2, "--config", config, "--target", "clean"]
+
+        _run_subprocess(cmd_args)
+
+
+def build_targets(args, cmake_path: Path, build_dir: Path, configs: Set[str], num_parallel_jobs: int):
+
+    env = {}
+    if args.android:
+        env["ANDROID_HOME"] = str(args.android_home)
+        env["ANDROID_NDK_HOME"] = str(args.android_ndk_path)
+
+    for config in configs:
+        log.info("Building targets for %s configuration", config)
+        build_dir2 = _get_build_config_dir(build_dir, config)
+        cmd_args = [str(cmake_path), "--build", str(build_dir2), "--config", config]
+
+        build_tool_args = []
+        if num_parallel_jobs != 1:
+            if is_windows() and args.cmake_generator != "Ninja" and not args.build_wasm:
+                build_tool_args += [
+                    "/maxcpucount:{}".format(num_parallel_jobs),
+                    # if nodeReuse is true, msbuild processes will stay around for a bit after the build completes
+                    "/nodeReuse:False",
+                ]
+            elif is_macOS() and args.use_xcode:
+                # CMake will generate correct build tool args for Xcode
+                cmd_args += ["--parallel", str(num_parallel_jobs)]
+            else:
+                build_tool_args += ["-j{}".format(num_parallel_jobs)]
+
+        if build_tool_args:
+            cmd_args += ["--"]
+            cmd_args += build_tool_args
+
+        _run_subprocess(cmd_args, env=env)
+
+
+def _run_python_tests():
+    # TODO: Run the python tests in /python
+    pass
+
+
+def _run_android_tests(args, build_dir: Path, config: str, cwd: Path):
+    # TODO: Setup running tests using Android simulator and adb. See ORT build.py for example.
+    source_dir = REPO_DIR
+    pass
+
+
+def _run_ios_tests(args, config: str, cwd: Path):
+    # TODO: Setup running tests using xcode an iPhone simulator. See ORT build.py for example.
+    source_dir = REPO_DIR
+    pass
+
+
+def _run_cxx_tests(args, build_dir: Path, configs: Set[str]):
+    ctest_path = None if not args.ctest_path else _resolve_executable_path(args.ctest_path)
+
+    for config in configs:
+        log.info("Running tests for %s configuration", config)
+
+        cwd = _get_build_config_dir(build_dir, config)
+
+        if args.android:
+            _run_android_tests(args, build_dir, config, cwd)
+            continue
+        elif args.ios:
+            _run_ios_tests(args, config, cwd)
+            continue
+
+        if ctest_path:
+            ctest_cmd = [str(ctest_path), "--build-config", config, "--verbose", "--timeout", "10800"]
+            _run_subprocess(ctest_cmd, cwd=cwd)
+
+        else:
```

Comment:
why do we have both a ctest and non-ctest way to run tests?

Reply:
To get code coverage from the tests we'd need to run manually. 

Maybe we should default to ctest (assume it's available in PATH if not explicitly provided) unless the user requests code coverage to be generated (windows only, execute test exes directly with vstest)

Reply:
I see. maybe a separate option for code coverage would be good. the default is to use "ctest" now. one has to specify (oddly) `--ctest_path=""` to not use ctest.

Reply:
Updated to use ctest unless code coverage (windows only) is requested. as the user has cmake it seems reasonable that ctest is available, and is preferred as it avoids manually specifying tests to run.